pkg/store/lazy_postings.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"context"
+	"math"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/index"
+	"golang.org/x/exp/slices"
+
+	"github.com/thanos-io/thanos/pkg/block/indexheader"
+)
+
+var emptyLazyPostings = &lazyExpandedPostings{postings: nil, matchers: nil}
+
+// lazyExpandedPostings contains expanded postings (series IDs). If lazy posting expansion is
+// enabled, it might contain matchers that can be lazily applied during series filtering time.
+type lazyExpandedPostings struct {
+	postings []storage.SeriesRef
+	matchers []*labels.Matcher
+}
+
+func newLazyExpandedPostings(ps []storage.SeriesRef, matchers ...*labels.Matcher) *lazyExpandedPostings {
+	return &lazyExpandedPostings{
+		postings: ps,
+		matchers: matchers,
+	}
+}
+
+func (p *lazyExpandedPostings) lazyExpanded() bool {
+	return p != nil && len(p.matchers) > 0
+}
+
+func optimizePostingsFetchByDownloadedBytes(r *bucketIndexReader, postingGroups []*postingGroup, seriesMaxSize int64, seriesMatchRatio float64) ([]*postingGroup, bool, error) {
+	if len(postingGroups) <= 1 {
+		return postingGroups, false, nil
+	}
+	// Collect posting cardinality of each posting group.
+	for _, pg := range postingGroups {
+		// A posting group can have either add keys or remove keys but not both the same time.
+		vals := pg.addKeys
+		if len(pg.removeKeys) > 0 {
+			vals = pg.removeKeys
+		}
+		rngs, err := r.block.indexHeaderReader.PostingsOffsets(pg.name, vals...)
+		if err != nil {
+			return nil, false, errors.Wrapf(err, "postings offsets for %s", pg.name)
+		}
+
+		// No posting ranges found means empty posting.
+		if len(rngs) == 0 {
+			return nil, true, nil
+		}
+		for _, r := range rngs {
+			if r == indexheader.NotFoundRange {
+				continue
+			}
+			pg.cardinality += (r.End - r.Start - 4) / 4
```

Comment:
`-4` because the postings length is encoded in the 4 bytes, right? Perhaps worth adding a comment here?

Reply:
Added a comment about the 4 bytes